test/integration/sapi-hmac-auth.int.c

```diff
@@ -0,0 +1,431 @@
+    }
+
+    /*
+     * Finally!!  Write the data to the NV index.
+     * If the command is successful, the command
+     * HMAC was correct.
+     */
+    rc = TSS2_RETRY_EXP (Tss2_Sys_NV_Write (sys_ctx,
+                            TPM20_INDEX_PASSWORD_TEST,
+                            TPM20_INDEX_PASSWORD_TEST,
+                            &nvCmdAuths,
+                            &nvWriteData,
+                            0,
+                            &nvRspAuths));
+    if (rc != TSS2_RC_SUCCESS) {
+        LOG_ERROR ("Tss2_Sys_NV_Write failed with RC: 0x%x", rc);
+        return rc;
+    }
+
+
+    /* Roll nonces for response */
+    roll_nonces (nvSession, &nvRspAuths.auths[0].nonce);
+
+    /*
+     * If the command was successful, check the
+     * response HMAC to make sure that the
+     * response was received correctly.
+     */
+    rc = check_response_hmac (sys_ctx,
+                              &nvCmdAuths,
+                              TPM20_INDEX_PASSWORD_TEST,
+                              TPM20_INDEX_PASSWORD_TEST,
+                              TPM2_RH_NULL,
+                              &nvRspAuths);
+    if (rc != TSS2_RC_SUCCESS) {
+        LOG_ERROR ("check_response_hmac failed with RC: 0x%x", rc);
+        return rc;
+    }
+
+    if (session_type == TPM2_SE_POLICY) {
+        rc = Tss2_Sys_PolicyAuthValue (sys_ctx,
+                                       nvSession->sessionHandle,
+                                       0,
+                                       0);
+        if (rc != TSS2_RC_SUCCESS) {
+            LOG_ERROR ("Tss2_Sys_PolicyAuthValue failed with RC: 0x%x", rc);
+            return rc;
+        }
+    }
+    /* First call prepare in order to create cpBuffer. */
+    rc = Tss2_Sys_NV_Read_Prepare (sys_ctx,
+                                   TPM20_INDEX_PASSWORD_TEST,
+                                   TPM20_INDEX_PASSWORD_TEST,
+                                   NV_DATA_SIZE,
+                                   0);
+    if (rc != TSS2_RC_SUCCESS) {
+        LOG_ERROR ("Tss2_Sys_NV_Read_Prepare failed with RC: 0x%x", rc);
+        return rc;
+    }
+
+    roll_nonces (nvSession, &nvCmdAuths.auths[0].nonce);
+    /* End the session after next command. */
+    nvCmdAuths.auths[0].sessionAttributes &= ~TPMA_SESSION_CONTINUESESSION;
+
+    /*
+     * Complete command authorization area, by computing
+     * HMAC and setting it in nvCmdAuths.
+     */
+    rc = compute_command_hmac (sys_ctx,
+                               TPM20_INDEX_PASSWORD_TEST,
+                               TPM20_INDEX_PASSWORD_TEST,
+                               TPM2_RH_NULL,
+                               &nvCmdAuths);
+    if (rc != TSS2_RC_SUCCESS) {
+        LOG_ERROR ("compute_command_hmac failed with RC: 0x%x", rc);
+        return rc;
+    }
+
+    /*
+     * And now read the data back.
+     * If the command is successful, the command
+     * HMAC was correct.
+     */
+    rc = Tss2_Sys_NV_Read (sys_ctx,
```

Comment:
This will call Tss2_Sys_NV_Read_Prepare() again. Shouldn't this rather be
Tss2_Sys_Execute() followed by Tss2_Sys_NV_Read_Complete()?
All the auths will probably be the same, only the params will be written twice.

Reply:
Are you waiting for me to make this change? Not clear since you've marked the PR 'approved'. I don't disagree with your analysis but the outcome is effectively the same.